File: swift/problem367.swift
//linear search
class Solution1 {
    func isPerfectSquare(_ num: Int) -> Bool {
        var root = 1
        
        while (root*root < num) {
            root += 1
        }
        
        return root*root == num
    }
}

//Newton's method
//refer to cpp solution
class Solution2 {
    func isPerfectSquare(_ num: Int) -> Bool {
        var x0 = 0; var x1 = num/2 + 1
        
        repeat {
            x0 = x1
            x1 = (num+x0*x0)/(2*x0)
            if (x1*x1 == num) {
                return true
            }
        } while (abs(x1-x0) > 1)
        
        return false
    }
}

File: java/problem367.java
class Solution {
    public boolean isPerfectSquare(int num) {
        int beg=0;int end=65536;//2**16
        int med;
        while(true){
            if(end-beg<=1){
                break;
            }
            med=(beg+end)/2;
            int temp=med*med;
            if(temp>num||temp<0){
                end=med;
            }else if(temp<num){
                beg=med;
            }else{
                return true;
            }
            
        }
        return beg*beg==num||end*end==num;
    }
}
public class problem367{
     public static void main(String args[]){
         Solution test=new Solution();
         System.out.println(test.isPerfectSquare(2147395600));
    }
}